lib/request.js

```diff
@@ -595,77 +591,28 @@ AWS.Request = inherit({
           resp.error = error;
           resp.error.retryable = false;
         });
-
-        var shouldCheckContentLength = false;
-        var expectedLen;
-        if (req.httpRequest.method !== 'HEAD') {
-          expectedLen = parseInt(headers['content-length'], 10);
-        }
-        if (expectedLen !== undefined && !isNaN(expectedLen) && expectedLen >= 0) {
-          shouldCheckContentLength = true;
-          var receivedLen = 0;
-        }
-
-        var checkContentLengthAndEmit = function checkContentLengthAndEmit() {
-          if (shouldCheckContentLength && receivedLen !== expectedLen) {
-            stream.emit('error', AWS.util.error(
-              new Error('Stream content length mismatch. Received ' +
-                receivedLen + ' of ' + expectedLen + ' bytes.'),
-              { code: 'StreamContentLengthMismatch' }
-            ));
-          } else if (AWS.HttpClient.streamsApiVersion === 2) {
-            stream.end();
-          } else {
-            stream.emit('end');
-          }
-        };
-
         var httpStream = resp.httpResponse.createUnbufferedStream();
-
         if (AWS.HttpClient.streamsApiVersion === 2) {
-          if (shouldCheckContentLength) {
-            var lengthAccumulator = new streams.PassThrough();
-            lengthAccumulator._write = function(chunk) {
-              if (chunk && chunk.length) {
-                receivedLen += chunk.length;
-              }
-              return streams.PassThrough.prototype._write.apply(this, arguments);
-            };
-
-            lengthAccumulator.on('end', checkContentLengthAndEmit);
-            stream.on('error', function(err) {
-              shouldCheckContentLength = false;
-              httpStream.unpipe(lengthAccumulator);
-              lengthAccumulator.emit('end');
-              lengthAccumulator.end();
-            });
-            httpStream.pipe(lengthAccumulator).pipe(stream, { end: false });
-          } else {
             httpStream.pipe(stream);
-          }
         } else {
-
-          if (shouldCheckContentLength) {
```

Comment:
We'll still need to perform these checks for Node.js 0.8, especially if we start turning it on by default. Unfortunately 0.8.x doesn't include `Transform` in its `stream` package, so your implementation won't work for those cases. You'll likely need to resort to `data` listeners to perform your calculations if `Transform` doesn't exist.